### ISSUE TITLE
Enable DiscoveryClient from all namespaces based on property opt in

### ISSUE THREAD

**Issue by Incarnation-p-lee** [spring-cloud/spring-cloud-kubernetes#428]
As we know, the package
```pom
<dependency>
    <groupId>org.springframework.cloud</groupId>
    <artifactId>spring-cloud-starter-kubernetes</artifactId>
</dependency>
```
will discovery all services from k8s, and we can configure the namespace from property
```
spring:
  cloud:
    kuberenetes:
      client:
        namespace: ${my-namespace}
```

But how can I discovery services from all namespace ?

**Comment by geoand** [comment]
Hi,

You could define a custom `KubernetesClientServicesFunction` bean that would look something like this:

```
@Configuration
public class KubernetesClientServicesFunctionConfiguration {
 
   @Bean
   public KubernetesClientServicesFunction servicesFunction() {
      return client.services().inAnyNamespace()
   }
}
```

**Comment by Incarnation-p-lee** [comment]
Thanks a lot. but what is this `client` here?

**Comment by geoand** [comment]
Good catch! I updated the answer

**Comment by Incarnation-p-lee** [comment]
seems need add `@Primary` for this bean, let me have a try again,

**Comment by Incarnation-p-lee** [comment]
seems not working, I cannot see any service now.

**Comment by Incarnation-p-lee** [comment]
looks some error like
`The bean 'servicesFunction', defined in class path resource [org/springframework/cloud/kubernetes/discovery/KubernetesDiscoveryClientAutoConfiguration.class], could not be registered. A bean with that name has already been defined in class path resource [com/microsoft/azure/spring/service/config/CloudGatewayConfiguration.class] and overriding is disabled.`

**Comment by geoand** [comment]
Can you then add `@Primary`? 

The reason I ask to check the above configuration is because that is essentially what happens by default. If you add it and continue to see no services then there is something else going on

**Comment by Incarnation-p-lee** [comment]
Yes, I added `@Primary`, it looks that bean added `@Primary` too.

**Comment by ryanjbaxter** [comment]
@Incarnation-p-lee can you open an issue so we can update the `servicesFunction` bean name to be more specific, it is too generic (obviously).

I think this would be a nice enhancement that we could add so based on a property we create the bean for you.

**Comment by Incarnation-p-lee** [comment]
how about `Support kubernetes discovery client discovery services from all namespace` as feature request ?

**Comment by spencergibb** [comment]
> I think this would be a nice enhancement that we could add so based on a property we create the bean for you.

This issue has been marked as an enhancement.

**Comment by ryanjbaxter** [comment]
Closed via 702c7b2
